mysqld_exporter.go

```diff
@@ -225,36 +241,47 @@ func (e *Exporter) setGenericMetric(name string, value float64) {
 	e.metrics[name].Set(value)
 }
 
-func (e *Exporter) setMetrics(scrapes <-chan []string) {
-	var comRegex = regexp.MustCompile(`^com_(.*)$`)
-	var connectionErrorRegex = regexp.MustCompile(`^connection_errors_(.*)$`)
-	var innodbRowRegex = regexp.MustCompile(`^innodb_rows_(.*)$`)
-	var performanceSchemaRegex = regexp.MustCompile(`^performance_schema_(.*)$`)
+var (
+	comRegex               = regexp.MustCompile(`^com_(.*)$`)
+	connectionErrorRegex   = regexp.MustCompile(`^connection_errors_(.*)$`)
+	innodbRowRegex         = regexp.MustCompile(`^innodb_rows_(.*)$`)
+	performanceSchemaRegex = regexp.MustCompile(`^performance_schema_(.*)$`)
+)
+
+func (e *Exporter) setGlobalStatusMetric(name string, value float64) {
+	command := comRegex.FindStringSubmatch(name)
+	connectionError := connectionErrorRegex.FindStringSubmatch(name)
+	innodbRow := innodbRowRegex.FindStringSubmatch(name)
+	performanceSchema := performanceSchemaRegex.FindStringSubmatch(name)
+
+	switch {
+	case len(command) > 0:
+		e.commands.With(prometheus.Labels{"command": command[1]}).Set(value)
+	case len(connectionError) > 0:
+		e.connectionErrors.With(prometheus.Labels{"error": connectionError[1]}).Set(value)
+	case len(innodbRow) > 0:
+		e.innodbRows.With(prometheus.Labels{"operation": innodbRow[1]}).Set(value)
+	case len(performanceSchema) > 0:
+		e.performanceSchema.With(prometheus.Labels{"instrumentation": performanceSchema[1]}).Set(value)
+	default:
+		e.setGenericMetric(globalStatus, name, value)
+	}
```

Comment:
Flow is a bit clunky here. Also, all regexp's are evaluation although we only need one hit.

How about the following:

if command := comRegex.FindStringSubmatch(name); command != nil {
    e.commands.With(prometheus.Labels{"command": command[1]}).Set(value)
    return
}
if //...
// and finally:
e.setGenericMetric(globalStatus, name, value)